// adequate-little-templates/dist/index.mjs
const truthy = (v) => !(v == null ||
    v === false ||
    v === 0 ||
    v === "" ||
    Number.isNaN(v) ||
    (Array.isArray(v) && v.length === 0) ||
    (typeof v === "object" &&
        v !== null &&
        !Array.isArray(v) &&
        Object.keys(v).length === 0));
const ck = (a, n, name) => a.length < n ? `[Error: ${name}() needs ${n} args]` : null;
const fns = {
    eq: (ctx, ...a) => ck(a, 2, "eq") ?? ev(a[0], ctx) === ev(a[1], ctx),
    ne: (ctx, ...a) => ck(a, 2, "ne") ?? ev(a[0], ctx) !== ev(a[1], ctx),
    gt: (ctx, ...a) => ck(a, 2, "gt") ?? Number(ev(a[0], ctx)) > Number(ev(a[1], ctx)),
    lt: (ctx, ...a) => ck(a, 2, "lt") ?? Number(ev(a[0], ctx)) < Number(ev(a[1], ctx)),
    gte: (ctx, ...a) => ck(a, 2, "gte") ?? Number(ev(a[0], ctx)) >= Number(ev(a[1], ctx)),
    lte: (ctx, ...a) => ck(a, 2, "lte") ?? Number(ev(a[0], ctx)) <= Number(ev(a[1], ctx)),
    and: (ctx, ...a) => {
        let r = true;
        for (const e of a) {
            r = ev(e, ctx);
            if (!truthy(r))
                return r;
        }
        return r;
    },
    or: (ctx, ...a) => {
        let r = false;
        for (const e of a) {
            r = ev(e, ctx);
            if (truthy(r))
                return r;
        }
        return r;
    },
    not: (ctx, ...a) => ck(a, 1, "not") ?? !truthy(ev(a[0], ctx)),
    lowercase: (ctx, ...a) => String(ev(a[0], ctx)).toLowerCase(),
    uppercase: (ctx, ...a) => String(ev(a[0], ctx)).toUpperCase(),
    trim: (ctx, ...a) => String(ev(a[0], ctx)).trim(),
    truncate: (ctx, ...a) => {
        const e = ck(a, 2, "truncate");
        if (e)
            return e;
        const s = String(ev(a[0], ctx)), n = Number(ev(a[1], ctx));
        const suffix = a[2] ? String(ev(a[2], ctx)) : "...";
        return s.length > n ? s.slice(0, n) + suffix : s;
    },
    replace: (ctx, ...a) => ck(a, 3, "replace") ??
        String(ev(a[0], ctx))
            .split(String(ev(a[1], ctx)))
            .join(String(ev(a[2], ctx))),
    limit: (ctx, ...a) => {
        const e = ck(a, 2, "limit");
        if (e)
            return e;
        const r = ev(a[0], ctx), n = ev(a[1], ctx);
        return Array.isArray(r) ? r.slice(0, n < 0 ? 0 : n) : r;
    },
    first: (ctx, ...a) => {
        const e = ck(a, 1, "first");
        if (e)
            return e;
        const r = ev(a[0], ctx);
        return Array.isArray(r) ? r[0] : r;
    },
    last: (ctx, ...a) => {
        const e = ck(a, 1, "last");
        if (e)
            return e;
        const r = ev(a[0], ctx);
        return Array.isArray(r) ? r[r.length - 1] : r;
    },
    length: (ctx, ...a) => {
        const e = ck(a, 1, "length");
        if (e)
            return e;
        const v = ev(a[0], ctx);
        return Array.isArray(v) ? v.length : String(v).length;
    },
    join: (ctx, ...a) => ck(a, 2, "join") ??
        ((r) => (Array.isArray(r) ? r.join(String(ev(a[1], ctx))) : String(r)))(ev(a[0], ctx)),
    default: (ctx, ...a) => {
        const e = ck(a, 2, "default");
        if (e)
            return e;
        const v = ev(a[0], ctx);
        return truthy(v) ? v : ev(a[1], ctx);
    },
    safeUrl: (ctx, ...a) => {
        const u = String(ev(a[0], ctx) ?? "").trim();
        return u && /^(?:\.{0,2}\/|[#?]|(?:https?|ftp):\/\/|(?:mailto|tel):)/i.test(u) ? u : "";
    },
};
const ev = (e, ctx) => {
    if (!e)
        return undefined;
    if (e.t === "L")
        return e.val;
    if (e.t === "V") {
        const has = (o, k) => Object.prototype.hasOwnProperty.call(o, k);
        let v = ctx;
        for (const k of e.path) {
            if (v == null || !has(v, k))
                return undefined;
            v = v[k];
        }
        return v;
    }
    const fn = fns[e.fn];
    if (!fn)
        return `[Error: unknown ${e.fn}()]`;
    return e.t === "C" ? fn(ctx, ...e.args) : fn(ctx, e.left, ...e.args);
};
const esc = (s) => s
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;")
    .replace(/'/g, "&#39;");
const rn = (nodes, ctx) => {
    let out = "";
    for (const n of nodes) {
        if (n.t === "T") {
            out += n.val;
            continue;
        }
        if (n.t === "I") {
            const v = ev(n.expr, ctx);
            if (Array.isArray(v))
                out += "[Error: use #each for arrays]";
            else if (typeof v === "object" && v !== null)
                out += "[Error: cannot render object]";
            else {
                const s = String(v ?? "");
                out += n.raw ? s : esc(s);
            }
            continue;
        }
        if (n.t === "F") {
            let matched = false;
            for (const b of n.branches)
                if (truthy(ev(b.cond, ctx))) {
                    out += rn(b.body, ctx);
                    matched = true;
                    break;
                }
            if (!matched && n.else)
                out += rn(n.else, ctx);
            continue;
        }
        if (n.t === "E") {
            const arr = ev(n.arr, ctx);
            if (!Array.isArray(arr)) {
                out += "[Error: #each needs array]";
                continue;
            }
            if (!arr.length && n.else)
                out += rn(n.else, ctx);
            else
                for (let i = 0; i < arr.length; i++) {
                    const local = { ...ctx, [n.as]: arr[i] };
                    if (n.idx)
                        local[n.idx] = i;
                    out += rn(n.body, local);
                }
        }
    }
    return out;
};
export const compile = (tmpl) => {
    let src = tmpl, pos = 0;
    const skipWs = () => {
        while (pos < src.length && " \t\n\r".includes(src[pos]))
            pos++;
    };
    const at = (s) => src.slice(pos, pos + s.length) === s;
    const skip = (s) => {
        if (at(s))
            pos += s.length;
    };
    const ident = () => {
        let r = "";
        while (pos < src.length && /\w/.test(src[pos]))
            r += src[pos++];
        return r;
    };
    const parseExpr = () => {
        skipWs();
        const start = pos;
        let expr;
        const ch = src[pos];
        if (ch === '"' || ch === "'") {
            const q = src[pos++];
            let s = "";
            while (pos < src.length && src[pos] !== q) {
                if (src[pos] === "\\" && pos + 1 < src.length) {
                    pos++;
                    const c = src[pos];
                    s += c === "n" ? "\n" : c === "t" ? "\t" : c === "r" ? "\r" : c;
                    pos++;
                    continue;
                }
                s += src[pos++];
            }
            if (pos < src.length)
                pos++;
            else
                s = "";
            expr = { t: "L", val: s };
        }
        else if (/[-0-9.]/.test(ch)) {
            let s = "", dot = 0;
            if (src[pos] === "-")
                s += src[pos++];
            if (src[pos] === ".") {
                s += src[pos++];
                dot = 1;
            }
            while (pos < src.length &&
                (/[0-9]/.test(src[pos]) || (src[pos] === "." && !dot++)))
                s += src[pos++];
            expr =
                s === "-" || s === "." || s === ""
                    ? { t: "V", path: [s || "-"] }
                    : { t: "L", val: parseFloat(s) };
        }
        else {
            const name = ident();
            if (name === "true")
                expr = { t: "L", val: true };
            else if (name === "false")
                expr = { t: "L", val: false };
            else if (name === "null")
                expr = { t: "L", val: null };
            else {
                skipWs();
                if (src[pos] === "(") {
                    pos++;
                    const args = [];
                    skipWs();
                    while (pos < src.length && src[pos] !== ")" && src[pos] !== "}") {
                        args.push(parseExpr());
                        skipWs();
                        if (src[pos] === ",") {
                            pos++;
                            skipWs();
                        }
                    }
                    if (src[pos] === ")")
                        pos++;
                    expr = { t: "C", fn: name, args };
                }
                else {
                    const path = [name];
                    while (src[pos] === ".") {
                        pos++;
                        path.push(ident());
                    }
                    expr = { t: "V", path };
                }
            }
        }
        skipWs();
        while (src[pos] === "|") {
            pos++;
            skipWs();
            const fn = ident();
            skipWs();
            const args = [];
            if (src[pos] === "(") {
                pos++;
                skipWs();
                while (pos < src.length && src[pos] !== ")" && src[pos] !== "}") {
                    args.push(parseExpr());
                    skipWs();
                    if (src[pos] === ",") {
                        pos++;
                        skipWs();
                    }
                }
                if (src[pos] === ")")
                    pos++;
            }
            expr = { t: "P", left: expr, fn, args };
            skipWs();
        }
        if (pos === start && pos < src.length)
            pos++;
        return expr;
    };
    const parseNodes = (stops = []) => {
        const result = [];
        outer: while (pos < src.length) {
            for (const s of stops)
                if (at(s))
                    break outer;
            if (src[pos] === "\\" && at("\\{{")) {
                pos++;
                result.push({ t: "T", val: "{{" });
                pos += 2;
                continue;
            }
            if (at("{{")) {
                pos += 2;
                skipWs();
                if (src[pos] === "+") {
                    pos++;
                    skipWs();
                    const expr = parseExpr();
                    skipWs();
                    skip("+");
                    skipWs();
                    while (pos < src.length && !at("}}"))
                        pos++;
                    skip("}}");
                    result.push({ t: "I", expr, raw: 1 });
                    continue;
                }
                if (src[pos] === "#") {
                    pos++;
                    const kw = ident();
                    skipWs();
                    if (kw === "if") {
                        const branches = [];
                        const cond = parseExpr();
                        skipWs();
                        skip("}}");
                        branches.push({
                            cond,
                            body: parseNodes([
                                "{{:else if",
                                "{{:elseif",
                                "{{:else}}",
                                "{{/if}}",
                            ]),
                        });
                        while (at("{{:else if") || at("{{:elseif")) {
                            pos += at("{{:elseif") ? 9 : 10;
                            skipWs();
                            const cond2 = parseExpr();
                            skipWs();
                            skip("}}");
                            branches.push({
                                cond: cond2,
                                body: parseNodes([
                                    "{{:else if",
                                    "{{:elseif",
                                    "{{:else}}",
                                    "{{/if}}",
                                ]),
                            });
                        }
                        let elseBody;
                        if (at("{{:else}}")) {
                            pos += 9;
                            elseBody = parseNodes(["{{/if}}"]);
                        }
                        skip("{{/if}}");
                        result.push({ t: "F", branches, else: elseBody });
                        continue;
                    }
                    if (kw === "each") {
                        const arrExpr = parseExpr();
                        skipWs();
                        const asKw = ident();
                        skipWs();
                        if (asKw !== "as") {
                            result.push({ t: "T", val: `[Error: #each missing 'as']` });
                            continue;
                        }
                        const itemName = ident();
                        skipWs();
                        let idxName;
                        if (src[pos] === ",") {
                            pos++;
                            skipWs();
                            idxName = ident();
                            skipWs();
                        }
                        skip("}}");
                        const body = parseNodes(["{{:else}}", "{{/each}}"]);
                        let elseBody;
                        if (at("{{:else}}")) {
                            pos += 9;
                            elseBody = parseNodes(["{{/each}}"]);
                        }
                        skip("{{/each}}");
                        result.push({
                            t: "E",
                            arr: arrExpr,
                            as: itemName,
                            idx: idxName,
                            body,
                            else: elseBody,
                        });
                        continue;
                    }
                    result.push({ t: "T", val: `[Error: unknown #${kw}]` });
                    skipWs();
                    skip("}}");
                    continue;
                }
                const expr = parseExpr();
                skipWs();
                while (pos < src.length && !at("}}"))
                    pos++;
                skip("}}");
                result.push({ t: "I", expr });
                continue;
            }
            let text = "";
            while (pos < src.length) {
                for (const s of stops)
                    if (at(s)) {
                        if (text)
                            result.push({ t: "T", val: text });
                        continue outer;
                    }
                if (src[pos] === "\\" && at("\\{{"))
                    break;
                if (at("{{"))
                    break;
                text += src[pos++];
            }
            if (text)
                result.push({ t: "T", val: text });
        }
        return result;
    };
    const ast = parseNodes();
    return (data) => rn(ast, data);
};
export const render = (tmpl, data) => compile(tmpl)(data);
export const registerFunction = (name, fn) => {
    fns[name] = (ctx, ...a) => fn(...a.map((e) => ev(e, ctx)));
};
